File: 백준/Silver/9012. 괄호/괄호.py
t = int(input())

for _ in range(t):
    g = input()
    stack = []

    for i in g:
        if i =='(':
           stack.append(i)
        elif i == ')':
            if stack:
                stack.pop()
            else:
                print("NO")
                break
    else:
        if not stack:
            print("YES")
        else:
            print("NO")

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());
        for (int i = 0; i < T; i++){
            Stack<Character> stack = new Stack<>();
            String vps = br.readLine();
            boolean isVPS = true;
            for(char c : vps.toCharArray()){
                if(c == '('){
                    stack.push(c);
                } else if(c == ')'){
                    if(stack.isEmpty()){
                        isVPS = false;
                        break;
                    }
                    stack.pop();
                }
            }

            if(isVPS && stack.isEmpty()){
                System.out.println("YES");
            }else{
                System.out.println("NO");
            }

        }
    }
}